VM to VM calls (#392)
powdr-labs/powdr: analysis/README.md

# Analysis

This crate is where most of the compilation pipeline lives. It takes a parse tree and returns a tree of constrained machines.

## Virtual and constrained machines

> | Machine               | Virtual | Constrained |

Reviewer: agree

Virtual machine reduction turns virtual machines into constrained machines. It has no effect on machines which are already constrained.

#### Inference

Inference infers the value of assignment registers.

#### Batcher

The batcher groups statements within each function into batches which can be executed in the same execution step.

#### ASM to PIL

ASM to PIL has two steps: ROM generation and ROM reduction to fixed columns

##### ROM generation

Rom generation generates a single ROM for each virtual machine using the following process:
- Find the maximum number of inputs among all functions. Introduce as many input registers. Do the same for outputs, introducing output registers
- Replace references to the function arguments by references to these input registers.
- Pad all return statements with zeroes up to the number of output registers.
- Inline all functions, adding a label before each one.
- Add an infinite loop also behind a label
- Enable non-deterministically jumping to one of these labels by setting `function_id` to the index of the label

This process introduces instructions and registers, and returns the modified machine along with the ROM.

> ##### ROM to fixed

Author: Right. Constrained.

## Pipeline

We go through the pipeline using the following program as an example:

```
machine Main {

    degree 16;

    DifferentSignatures sub;

    reg pc[@pc];
    reg X[<=];
    reg Y[<=];

    instr identity X -> Y = sub.identity
    instr one -> Y = sub.one
    instr nothing = sub.nothing

    function main {
        nothing;
        return;
    }
}

// A machine exposing functions of different signatures
machine DifferentSignatures {

    reg pc[@pc];

    function identity x: field -> field {
        return x;
    }

    function one -> field {
        return 1;
    }

    function nothing {
        return;
    }
}
```

### Type checking

Type checking takes a parse tree and returns a tree of machines. The output type aims at being as strict as possible.

### Virtual machine reduction

Virtual machine reduction turns virtual machines into constrained machines. It has no effect on contrained machines.

#### Inference

Inference infers the value of assignment registers.

```diff
-               A <== one();
+               A <=Y= one();
```

#### Batcher

> The batcher groups statements within each function into batches which can be executed in the same execution step. In our example, we gain 33% in the main machine by batching the first label with the following statement.

Reviewer: which following statement?

Author: Fixed, didnt have the latest version of the example in here

Reviewer: I still don't get it. The next diff is still the END BATCH comments only.

Author: Yes, the batcher only inserts the batch delimitations. It does not change the statements.

Reviewer: Ah, I think I misunderstood the text. So "the following statement" refers to the statement after said label, right? I imagined a statement after the sentence

```diff
+               // END BATCH
+               // END BATCH
+               // END BATCH
+               // END BATCH Unimplemented
+               // END BATCH
```

#### ASM to PIL

ASM to PIL has two steps: ROM generation and reduction to constrained.

##### ROM generation

Rom generation generates a single ROM for each virtual machine using the following process:
- Find the maximum number of inputs among all functions. Introduce as many input registers. Do the same for outputs, introducing output registers
- Replace references to the function arguments by references to these input registers.
- Pad all return statements with zeroes up to the number of output registers.
- Inline all functions, adding a label before each one.
- Add an infinite loop also behind a label
- Enable non-deterministically jumping to one of these labels by setting `function_id` to the index of the label

This process introduces instructions and registers, and returns the modified machine along with the ROM.

> 

Reviewer: Can you also add an example of rom gen with the dispatcher?

Author: Added the two roms for the example

Reviewer: nice

+               pol constant p_line = [0, 1, 2, 3, 4, 5] + [5]*;
+               pol commit _output_0_free_value(i) query match pc {  };
+               pol constant p__output_0_const = [0, 0, 0, 1, 0, 0] + [0]*;
+               pol constant p__output_0_read_free = [0, 0, 0, 0, 0, 0] + [0]*;
+               pol constant p_instr__jump_to_operation = [0, 1, 0, 0, 0, 0] + [0]*;
+               pol constant p_instr__loop = [0, 0, 0, 0, 0, 1] + [1]*;
+               pol constant p_instr__reset = [1, 0, 0, 0, 0, 0] + [0]*;
+               pol constant p_instr_return = [0, 0, 1, 1, 1, 0] + [0]*;
+               pol constant p_read__output_0__input_0 = [0, 0, 1, 0, 0, 0] + [0]*;
+               pol constant p_read__output_0_pc = [0, 0, 0, 0, 0, 0] + [0]*;
+               { pc, instr__jump_to_operation, instr__reset, instr__loop, instr_return, _output_0_const, _output_0_read_free, read__output_0_pc, read__output_0__input_0 } in { p_line, p_instr__jump_to_operation, p_instr__reset, p_instr__loop, p_instr_return, p__output_0_const, p__output_0_read_free, p_read__output_0_pc, p_read__output_0__input_0 };
-       function nothing {
-               return;
-               // END BATCH
+

> // the body of functions is removed and they are assigned their function id. Their inputs and outputs are mapped to the relevant registers

Reviewer: ```suggestion
// function bodies are removed and functions are assigned their function id. Their inputs and outputs are mapped to the relevant registers
```

Reviewer: but this hasn't been done here yet, right?

Reviewer: ah sorry, they are

+               pol constant p_read_X_A = [0, 0, 0, 0, 0] + [0]*;
+               pol constant p_read_X_pc = [0, 0, 0, 0, 0] + [0]*;
+               pol constant p_read_Y_A = [0, 0, 0, 0, 0] + [0]*;
+               pol constant p_read_Y_pc = [0, 0, 0, 0, 0] + [0]*;
+               pol constant p_reg_write_X_A = [0, 0, 0, 0, 0] + [0]*;
+               pol constant p_reg_write_Y_A = [0, 0, 1, 0, 0] + [0]*;
+               { pc, reg_write_X_A, reg_write_Y_A, instr_identity, instr_one, instr_nothing, instr__jump_to_operation, instr__reset, instr__loop, instr_return, X_const, X_read_free, read_X_A, read_X_pc, Y_const, Y_read_free, read_Y_A, read_Y_pc } in { p_line, p_reg_write_X_A, p_reg_write_Y_A, p_instr_identity, p_instr_one, p_instr_nothing, p_instr__jump_to_operation, p_instr__reset, p_instr__loop, p_instr_return, p_X_const, p_X_read_free, p_read_X_A, p_read_X_pc, p_Y_const, p_Y_read_free, p_read_Y_A, p_read_Y_pc };
+       }
+
-       function main {
-               start::
-               A <=Y= one();
-               // END BATCH Unimplemented
-               return;
-               // END BATCH
+       function main<2> {
+
```

> 

Reviewer: Nice!

### Block enforcer

This step takes constrained machines and enforces that the `function_id` can only change if the `latch` is on. This defines blocks of computation which can be created based on the functions exposed by each machine.

We add an identical block of constraints for each machine type:
```diff
+       constraints {
+               pol constant _block_enforcer_last_step = [0]* + [1];
+               pol commit _function_id_no_change;
+               _function_id_no_change = ((1 - _block_enforcer_last_step) * (1 - instr_return));
+               (_function_id_no_change * (_function_id' - _function_id)) = 0;
+       }
+
+       }
+
+       constraints {
+               pol constant _block_enforcer_last_step = [0]* + [1];
+               pol commit _function_id_no_change;
+               _function_id_no_change = ((1 - _block_enforcer_last_step) * (1 - instr_return));
+               (_function_id_no_change * (_function_id' - _function_id)) = 0;
```

> 

Reviewer: This looks a bit off. The `}` on line 265 is extra and the `{` online 269 isn't closed. Also these are identical, as explained above. But how does the compiler treat this?

Author: I adjusted the display for the diff to be clearer. I don't understand the last question?

Reviewer: I think I get it now. Each of these identical blocks are part of different machines, so there is no conflict. Maybe it makes sense to just show one block here (instead of 2) and just say each machine gets one of them (which is already said).

### Airgen

> Airgen takes machines which are only left with constraints and external instructions and instanciates them as a tree of AIR objects. Objects can point to each other using links, which encode the interaction between different machines.

Reviewer: ```suggestion
Airgen takes machines which are only left with constraints and external instructions and instantiates them as a tree of AIR objects. Objects can point to each other using links, which encode the interaction between different machines.
```

                pol constant p_read_Y_A = [0, 0, 0, 0, 0] + [0]*;
                pol constant p_read_Y_pc = [0, 0, 0, 0, 0] + [0]*;
                pol constant p_reg_write_X_A = [0, 0, 0, 0, 0] + [0]*;
                pol constant p_reg_write_Y_A = [0, 0, 1, 0, 0] + [0]*;
                { pc, reg_write_X_A, reg_write_Y_A, instr_identity, instr_one, instr_nothing, instr__jump_to_operation, instr__reset, instr__loop, instr_return, X_const, X_read_free, read_X_A, read_X_pc, Y_const, Y_read_free, read_Y_A, read_Y_pc } in { p_line, p_reg_write_X_A, p_reg_write_Y_A, p_instr_identity, p_instr_one, p_instr_nothing, p_instr__jump_to_operation, p_instr__reset, p_instr__loop, p_instr_return, p_X_const, p_X_read_free, p_read_X_A, p_read_X_pc, p_Y_const, p_Y_read_free, p_read_Y_A, p_read_Y_pc };
        }

        constraints {
                pol constant _block_enforcer_last_step = [0]* + [1];
                pol commit _function_id_no_change;
                _function_id_no_change = ((1 - _block_enforcer_last_step) * (1 - instr_return));
                (_function_id_no_change * (_function_id' - _function_id)) = 0;
        }

>         instr identity X -> Y { sub.identity }

Reviewer: This syntax looks kinda weird, as if `sub.identity` needs to be a true constraint. What happened to the previous `instr identity X -> Y = sub.identity;`?

Author: This was an issue in Display. Fixed.

pol constant p_instr_one = [0, 0, 1, 0, 0] + [0]*;
pol constant p_instr_return = [0, 0, 0, 1, 0] + [0]*;
pol constant p_read_X_A = [0, 0, 0, 0, 0] + [0]*;
pol constant p_read_X_pc = [0, 0, 0, 0, 0] + [0]*;
pol constant p_read_Y_A = [0, 0, 0, 0, 0] + [0]*;
pol constant p_read_Y_pc = [0, 0, 0, 0, 0] + [0]*;
pol constant p_reg_write_X_A = [0, 0, 0, 0, 0] + [0]*;
pol constant p_reg_write_Y_A = [0, 0, 1, 0, 0] + [0]*;
{ pc, reg_write_X_A, reg_write_Y_A, instr_identity, instr_one, instr_nothing, instr__jump_to_operation, instr__reset, instr__loop, instr_return, X_const, X_read_free, read_X_A, read_X_pc, Y_const, Y_read_free, read_Y_A, read_Y_pc } in { p_line, p_reg_write_X_A, p_reg_write_Y_A, p_instr_identity, p_instr_one, p_instr_nothing, p_instr__jump_to_operation, p_instr__reset, p_instr__loop, p_instr_return, p_X_const, p_X_read_free, p_read_X_A, p_read_X_pc, p_Y_const, p_Y_read_free, p_read_Y_A, p_read_Y_pc };
pol constant _block_enforcer_last_step = [0]* + [1];
pol commit _function_id_no_change;
_function_id_no_change = ((1 - _block_enforcer_last_step) * (1 - instr_return));
(_function_id_no_change * (_function_id' - _function_id)) = 0;
// Links:

> // // // instr identity with params  X -> Y links to // // function identity with id 2 with params  _input_0 -> _output_0 in object at location main_sub with latch "instr_return" and function_id "_function_id"

Reviewer: Not too important but redundant `//` here and later in the same line

Reviewer: also, where are these links?

Author: Fixed. What do you mean where? They belong to the object.

Reviewer: yes but at first I thought the comments were referring to something right underneath or something close

pol constant p_line = [0, 1, 2, 3, 4, 5] + [5]*;
pol commit _output_0_free_value(i) query match pc {  };
pol constant p__output_0_const = [0, 0, 0, 1, 0, 0] + [0]*;
pol constant p__output_0_read_free = [0, 0, 0, 0, 0, 0] + [0]*;
pol constant p_instr__jump_to_operation = [0, 1, 0, 0, 0, 0] + [0]*;
pol constant p_instr__loop = [0, 0, 0, 0, 0, 1] + [1]*;
pol constant p_instr__reset = [1, 0, 0, 0, 0, 0] + [0]*;
pol constant p_instr_return = [0, 0, 1, 1, 1, 0] + [0]*;
pol constant p_read__output_0__input_0 = [0, 0, 1, 0, 0, 0] + [0]*;
pol constant p_read__output_0_pc = [0, 0, 0, 0, 0, 0] + [0]*;
{ pc, instr__jump_to_operation, instr__reset, instr__loop, instr_return, _output_0_const, _output_0_read_free, read__output_0_pc, read__output_0__input_0 } in { p_line, p_instr__jump_to_operation, p_instr__reset, p_instr__loop, p_instr_return, p__output_0_const, p__output_0_read_free, p_read__output_0_pc, p_read__output_0__input_0 };
pol constant _block_enforcer_last_step = [0]* + [1];
pol commit _function_id_no_change;
_function_id_no_change = ((1 - _block_enforcer_last_step) * (1 - instr_return));
(_function_id_no_change * (_function_id' - _function_id)) = 0;
```

### Linker

> The linker takes a tree of machines and instanciates specific structures to make proofs about. In our current implementation, it simply collates all AIR objects next to each other in a single table.

Reviewer: ```suggestion
The linker takes a tree of machines and instantiates specific structures that will be used in proof generation. In our current implementation, it simply collates all AIR objects next to each other in a single table.
```
